onnxruntime/wasm/api.cc

```diff
@@ -387,3 +395,109 @@ char* OrtEndProfiling(ort_session_handle_t session) {
+}
+
+int EMSCRIPTEN_KEEPALIVE OrtTrainingLazyResetGrad(ort_training_session_handle_t training_handle) {
+  return CHECK_TRAINING_STATUS(LazyResetGrad, training_handle);
+}
+
+OrtValue* EMSCRIPTEN_KEEPALIVE OrtRunTrainStep(ort_training_session_handle_t training_handle,
+                                               const ort_run_options_handle_t options,
+                                               const size_t inputs_len,
+                                               const ort_tensor_handle_t* inputs,
+                                               const size_t outputs_len,
+                                               ort_tensor_handle_t* outputs) {
+  return (CHECK_TRAINING_STATUS(TrainStep, training_handle,
+                                options, inputs_len, inputs, outputs_len, outputs) == ORT_OK)
+             ? *outputs
+             : nullptr;
+}
+
+int EMSCRIPTEN_KEEPALIVE OrtTrainingOptimizerStep(ort_training_session_handle_t training_handle,
+                                                  const ort_run_options_handle_t run_options) {
+  return CHECK_TRAINING_STATUS(OptimizerStep, training_handle, run_options);
+}
+
+ort_tensor_handle_t EMSCRIPTEN_KEEPALIVE OrtTrainingEvalStep(ort_training_session_handle_t training_handle,
+                                                             const ort_run_options_handle_t options,
+                                                             size_t inputs_len,
+                                                             const ort_tensor_handle_t* inputs,
+                                                             size_t outputs_len,
+                                                             ort_tensor_handle_t* outputs) {
+  return (CHECK_TRAINING_STATUS(EvalStep, training_handle,
+                                options, inputs_len, inputs, outputs_len, outputs) == ORT_OK)
+             ? *outputs
+             : nullptr;
+}
+
+// void EMSCRIPTEN_KEEPALIVE OrtTrainingExportModelForInferencing(orttraining_session_handle_t session,
+//                                                                const ORTCHAR_T* inference_model_path,
+//                                                                size_t graph_outputs_len,
+//                                                                const char* const* graph_output_names) {
+//   Ort::GetTrainingApi().ExportModelForInferencing(session, inference_model_path, graph_outputs_len, graph_output_names);
+// }
+
+size_t* EMSCRIPTEN_KEEPALIVE OrtTrainingGetParametersSize(ort_training_session_handle_t training_handle,
+                                                          bool trainable_only) {
+  size_t* param_size = nullptr;
+  return (CHECK_TRAINING_STATUS(GetParametersSize, training_handle, param_size, trainable_only) == ORT_OK)
+             ? param_size
+             : nullptr;
+}
+
+ort_tensor_handle_t EMSCRIPTEN_KEEPALIVE OrtTrainingCopyParametersToBuffer(ort_training_session_handle_t training_handle,
```

Comment:
The expectation for this api is that user first calls the GetParameterSize api and then allocates a buffer of that size and then passes a pointer to this buffer with this API to get all the parameters in a contiguous buffer. (Used in FL scenarios). Will we do the same in web bindings? Will we allocate the buffer accessible to native code in JS side and then pass a pointer or do we need to do this here in this wrapper?

Reply:
JS will be able to allocate buffers of specific size through the webasembly (there's a malloc JS function, which is used for inference methods like createSession & run) -- whether we do it on the TS side or in the wrapper is up to how we want to design it, I think. Maybe we can have the alloc etc. be handled on the JS side for now, since it follows how inference handles buffers.

Reply:
should not return `ort_tensor_handle_t`

Reply:
and parameters_buffer should not be type of `ort_tensor_handle_t`. please check the type.

Reply:
please also fix the other functions